apps/emqx_gateway/src/coap/emqx_coap_channel.erl

```diff
@@ -0,0 +1,409 @@
+-define(DISCONNECT_WAIT_TIME, timer:seconds(10)).
+-define(INFO_KEYS, [conninfo, conn_state, clientinfo, session]).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+info(Channel) ->
+    maps:from_list(info(?INFO_KEYS, Channel)).
+
+info(Keys, Channel) when is_list(Keys) ->
+    [{Key, info(Key, Channel)} || Key <- Keys];
+
+info(conninfo, #channel{conninfo = ConnInfo}) ->
+    ConnInfo;
+info(conn_state, _) ->
+    connected;
+info(clientinfo, #channel{clientinfo = ClientInfo}) ->
+    ClientInfo;
+info(session, #channel{session = Session}) ->
+    emqx_misc:maybe_apply(fun emqx_session:info/1, Session);
+info(clientid, #channel{clientinfo = #{clientid := ClientId}}) ->
+    ClientId;
+info(ctx, #channel{ctx = Ctx}) ->
+    Ctx;
+info(subscribes, Channel) ->
+    Channel#channel.subscribes;
+info(tm, Channel) ->
+    Channel#channel.tm;
+info(resource, #channel{config = Cfg}) ->
+    maps:get(resource, Cfg);
+info(notify_type, #channel{config = Cfg}) ->
+    maps:get(notify_type, Cfg);
+info(subscribe_qos, #channel{config = Cfg}) ->
+    maps:get(subscribe_qos, Cfg);
+info(publish_qos, #channel{config = Cfg}) ->
+    maps:get(publish_qos, Cfg).
+
+stats(#channel{session = Session})->
+    emqx_session:stats(Session).
+
+init(ConnInfo = #{peername := {PeerHost, _},
+                  sockname := {_, SockPort}},
+     #{ctx := Ctx,
+       config := Config} = Option) ->
+    Peercert = maps:get(peercert, ConnInfo, undefined),
+    Mountpoint = maps:get(mountpoint, Option, undefined),
+    ClientInfo = set_peercert_infos(
+                   Peercert,
+                   #{ zone => default
+                    , protocol => 'mqtt-coap'
+                    , peerhost => PeerHost
+                    , sockport => SockPort
+                    , clientid => emqx_guid:to_base62(emqx_guid:gen())
+                    , username => undefined
+                    , is_bridge => false
+                    , is_superuser => false
+                    , mountpoint => Mountpoint
+                    }
+                  ),
+
+	#channel{ ctx = Ctx
+            , conninfo = ConnInfo
+            , clientinfo = ClientInfo
+            , tm = emqx_coap_tm:new()
+            , timers = #{}
+            , subscribes = emqx_coap_observe_res:new()
+            , config = Config
+            , keepalive = emqx_keepalive:init(maps:get(heartbeat, Config))
+            }.
+
+auth_publish(#channel{ctx = Ctx,
+                      clientinfo = ClientInfo},
+             Topic) ->
+    emqx_gateway_ctx:authorize(Ctx, ClientInfo, publish, Topic).
+
+auth_subscribe(#channel{ctx = Ctx,
+                        clientinfo = ClientInfo},
+               Topic) ->
+    emqx_gateway_ctx:authorize(Ctx, ClientInfo, subscribe, Topic).
+
+update(tm, Channel, TM) ->
+    Channel#channel{tm = TM};
+
+update(subscribes, #channel{subscribes = Manager} = Channel, Register) ->
```

Comment:
This is a better sequence of input parameters:
```
update(tm, TMValue, Channel) 
```

For example: `maps:update/3` is also like this.